### ISSUE TITLE
render text / plain text in controller response?

### ISSUE THREAD

**Issue by sam0x17** [amberframework/amber#306]
What is the amber equivalent of `render text: 'asfasdf'` (or in newer rails: `render body: 'asdfasdf'`)? I need to render raw text as part of an API.

**Comment by elorest** [comment]
```cr
def text_api
  "hello world!"
end
```

**Comment by sam0x17** [comment]
omg why didn't rails do it this way so simple!! :+1:

**Comment by elorest** [comment]
Yeah I love rails but this is definitely one of the things I wanted to do differently.

**Comment by c910335** [comment]
And don't forget `response.content_type = "test/plain"` or other Content-Type you need.

**Comment by sam0x17** [comment]
@c910335 is there a way I can do that in my base ApiController or do I need to do it in every single method. Also would super appreciate a JSON example?

**Comment by c910335** [comment]
[Custom Pipe](https://amberframework.org/guides#getting-started/routing/pipelines.md) or [Filters](https://amberframework.org/guides#getting-started/controllers/filters.md) should be good for you.

**Comment by elorest** [comment]
Yeah. Use just write a simple custom pipe to do that.

**Comment by elorest** [comment]
For JSON just write 
```cr
def api_action
  {"name" => "elorest"}.to_json
end
```

Also make sure to set the response type somehow.

**Comment by sam0x17** [comment]
excellent, thanks all

**Comment by elorest** [comment]
@sam0x17 Here's an example handler/pipe you that would set content_types for you.

```cr
    class TextType
      include HTTP::Handler
      def call(context : HTTP::Server::Context)
        content.response.content_type = "test/plain"
        call_next context
      end
    end
```

**Comment by sam0x17** [comment]
@elorest thanks!

**Comment by watzon** [comment]
Don't know how anyone else feels about this, but for purposes of readability and ease of use it could be nice to have a similar functionality to Rails here by using the render method with a named `json` attribute. That way the render method could take the hash, variable, etc, apply `to_json` if necessary, set the content type, and render the json response.

**Comment by drujensen** [comment]
@watzon something like this?  

https://github.com/kemalyst/kemalyst/blob/master/src/kemalyst/helpers.cr#L36

**Comment by watzon** [comment]
@drujensen something like that would work. Did that code not get ported over when kemalyst became amber?

**Comment by drujensen** [comment]
correct, Amber and Kemalyst had different approaches on this and it's been in discussion if we should port this more rails'ish style over or have the simpler style where you just return what you want as the body and use a handler or before filter to specify the `content-type`.

**Comment by watzon** [comment]
[insert "why not both" gif here]

I feel like it would be really nice to have the ease of using a helper function like that, but also have the ability to just return whatever I want. Maybe I just want to have my cake and eat it too haha.

**Comment by elorest** [comment]
My suggestion would be that we add them but don't require them.

```cr
class Amber::Controller::Base
  protected def text(body, status_code = 200)
    context.response.status_code = {{ status_code }}
    context.response.content_type = "text/plain"
    # We need to use content instead of response.print so that we don't close down the response object before session are written, in chunking mode.
    context.content = body
  end
end
```

Then we can modify the controller proc to not set content from the return value if it's already been set.

@drujensen WDYT?

**Comment by drujensen** [comment]
@elorest great idea.

**Comment by elorest** [comment]
I've began work on it. https://github.com/amberframework/amber/pull/313

**Comment by iambudi** [comment]
For Web/REST API currently i create a class with method helper to handle http response.
Each controller api then inherits that class. It's inspired from aspnet API methods.
I set all response in json or plain text.

```ruby
def response(status_code : Int32 = 200, content = "")
    # @response.headers["Content-Type"] = "text/plain"
    @context.response.content_type = "application/json"
    @context.response.status_code = status_code
    @context.content = to_json(content)
  end

  private def to_json(content : String | NamedTuple)
    content.is_a?(NamedTuple) ? content.to_json : content
  end
  
  macro ok(content = "")
    return response 200, {{content}}
  end

  # Other macros for created, updated, not_found etc
```

So when i need to send plain 200 response, simply i just write `ok("Success")`

**Comment by eliasjpr** [comment]
Resolved with #367
